src/Pact/Types/RowData.hs

```diff
@@ -0,0 +1,115 @@
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+-- |
+-- Module      :  Pact.Types.RowData
+-- Copyright   :  (C) 2021 Stuart Popejoy
+-- License     :  BSD-style (see the file LICENSE)
+-- Maintainer  :  Stuart Popejoy <stuart@kadena.io>
+--
+-- Versioned persistence for 'PactValue'.
+--
+module Pact.Types.RowData
+  ( RowData(..)
+  , RowDataVersion(..)
+  , RowDataValue(..)
+  , pactValueToRowData
+  , rowDataToPactValue
+  ) where
+
+import Control.Applicative
+import Control.DeepSeq (NFData)
+import Data.Aeson
+import Data.Default
+import Data.Text (Text)
+import Data.Vector (Vector)
+import GHC.Generics
+import Test.QuickCheck
+
+import Pact.Types.Exp
+import Pact.Types.PactValue
+import Pact.Types.Pretty
+import Pact.Types.Term
+
+data RowDataValue
+    = RDLiteral Literal
+    | RDList (Vector RowData)
+    | RDObject (ObjectMap RowData)
+    | RDGuard (Guard RowData)
+    | RDModRef ModRef
+    deriving (Eq,Show,Generic,Ord)
+instance NFData RowDataValue
+
+data RowDataVersion = RDV0 | RDV1
+  deriving (Eq,Show,Generic,Ord,Enum,Bounded)
+instance NFData RowDataVersion
+
+data RowData = RowData RowDataVersion RowDataValue
+  deriving (Eq,Show,Generic,Ord)
+instance NFData RowData
+instance Arbitrary RowData where
+  arbitrary = pactValueToRowData <$> arbitraryBoundedEnum <*> arbitrary
+
+pactValueToRowData :: RowDataVersion -> PactValue -> RowData
+pactValueToRowData v pv = RowData v $ case pv of
+  PLiteral l -> RDLiteral l
+  PList l -> RDList $ recur l
+  PObject o -> RDObject $ recur o
+  PGuard g -> RDGuard $ recur g
+  PModRef m -> RDModRef m
+  where
+    recur :: Functor f => f PactValue -> f RowData
+    recur = fmap $ pactValueToRowData v
+
+rowDataToPactValue :: RowData -> PactValue
+rowDataToPactValue (RowData _ rdv) = case rdv of
+  RDLiteral l -> PLiteral l
+  RDList l -> PList $ recur l
+  RDObject o -> PObject $ recur o
+  RDGuard g -> PGuard $ recur g
+  RDModRef m -> PModRef m
+  where
+    recur :: Functor f => f RowData -> f PactValue
+    recur = fmap $ rowDataToPactValue
+
+instance Pretty RowData where
+  pretty = pretty . rowDataToPactValue
+
+
+instance ToJSON RowData where
```

Comment:
I think in this instance, it makes sense to actually add `.version` or some sort of information to the json to it roundtrips properly.

`.version` (or whatever you want to call the field) if not present indicates `V0`, otherwise we match on the serialized `RDV1` and so forth.

Reply:
It's a good idea, maybe better at row level though to save on storage (otherwise it would show up for every nested value)

Reply:
Added toplevel version support with a proper `RowData` type